### ISSUE TITLE
Cloud-J in GCHP v14.3.0 crashed with Intel/Intelmpi environment

### ISSUE THREAD

**Issue by yuanjianz** [geoschem/gchp#387]
### Name and Institution (Required)

Name: Yuanjian Zhang
Institution: Washington University in St. Louis

### Confirm you have reviewed the following documentation

- [x] [Support guidelines](https://gchp.readthedocs.io/en/latest/reference/SUPPORT.html)
- [x] [User manuals](https://gchp.readthedocs.io/en/latest/geos-chem-shared-docs/supplemental-guides/related-docs.html)
- [x] [Debugging GEOS-Chem and HEMCO errors](https://gchp.readthedocs.io/en/latest/geos-chem-shared-docs/supplemental-guides/debug-guide.html)
- [x] [Current and past Github issues](https://github.com/geoschem/GCHP/issues)

### Description of your issue or question
In order to find a workaround for #385 , I tried to use another docker environment with Intel compiler and Intelmpi. The running pass through the hanging issue but soon crashed under another seg fault(see below and [lsf-687733.txt.log](https://github.com/geoschem/GCHP/files/14427271/lsf-687733.txt.log))
```
forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source             
gchp               00000000035C409A  Unknown               Unknown  Unknown
libpthread-2.17.s  00007F818914B630  Unknown               Unknown  Unknown
gchp               00000000016A2BA0  fjx_sub_mod_mp_bl        1355  fjx_sub_mod.f90
gchp               00000000016A2A5E  fjx_sub_mod_mp_mi        1303  fjx_sub_mod.f90
gchp               00000000016A1F99  fjx_sub_mod_mp_op        1214  fjx_sub_mod.f90
gchp               0000000001692EBC  fjx_sub_mod_mp_ph         572  fjx_sub_mod.f90
gchp               000000000168A127  cld_sub_mod_mp_cl         216  cld_sub_mod.f90
...
```
My environment is `1dandan/gchp-esmf8.3:lastest` with intel/intelmpi and ESMFv8.3.1. And I am using GCHP v14.3.0 /dev/no-diff-to-benchmark branch with GEOS-IT native C180 mass flux and default standard full chem. Resources requested are 60 cores for C30 simulation.
#### Issue Identification So Far
- /main branch with correct path for ExtData does not work
- another version of intel/intelmpi environment `1dandan/gchp-esmf8.3:2023-04` does not work either
- fjx_sub_mod.f90:1355 seems to be a common loop
```
1355       do K = 1,W_+W_r
1356       if (LDOKR(K) .gt. 0) then
1357        call GEN_ID (POMEGA(1,1,K),FZ(1,K),ZTAU(1,K),FSBOT(K),RFL(1,K), &
1358              PM,PM0, B(1,1,1,K),CC(1,1,1,K),AA(1,1,1,K), &
1359                      A(1,1,K),H(1,1,K),C(1,1,K), ND)
1360       endif
1361       enddo
```
- Using less resource identical to this issue (60 cores and C30) to overcome the hanging issue in #385 under gnu/openmpi successfully finished the running (So maybe this is a intel/intelmpi-only problem)
- I am currently trying to disable Cloud-J and enable Fast-JX and see if it fixs the problem

[gchp.20190701_0000z.log](https://github.com/geoschem/GCHP/files/14427270/gchp.20190701_0000z.log)
[lsf-687733.txt.log](https://github.com/geoschem/GCHP/files/14427271/lsf-687733.txt.log)
[warnings_and_errors.log](https://github.com/geoschem/GCHP/files/14427272/warnings_and_errors.log)
[allPEs.log](https://github.com/geoschem/GCHP/files/14427273/allPEs.log)

#### Update
Replacing Cloud-J with Fast-Jx does not crash and successfully produces results. So it should be a problem of Cloud-J under intel/intelmpi environment.

**Comment by lizziel** [comment]
Hi @yuanjianz, thanks for reporting this. Indeed this seems like an intel issue, and probably for a specific version since I have successfully run 14.3 with ifort before. Have you tried building with option -DCMAKE_BUILD_TYPE=Debug? If not, try doing that and rerunning and see if it you get more information in the log.

**Comment by yuanjianz** [comment]
Hi @lizziel, I tried building with Debug mode and then crashed at another subroutine.
```
forrtl: severe (61): format/variable-type mismatch, unit 6, file /proc/9/fd/1
Image              PC                Routine            Line        Source             
gchp               000000000AE91D1B  Unknown               Unknown  Unknown
gchp               000000000AEF4A5A  Unknown               Unknown  Unknown
gchp               00000000008D50C7  input_mod_mp_conf        2914  input_mod.F90
gchp               00000000008B0F80  input_mod_mp_read         222  input_mod.F90
gchp               0000000000569876  gchp_chunk_mod_mp         210  gchp_chunk_mod.F90
gchp               0000000000494A86  chem_gridcompmod_        1460  Chem_GridCompMod.F90
gchp               0000000009EC421E  Unknown               Unknown  Unknown
...
```
[gchp.20190701_0000z.log](https://github.com/geoschem/GCHP/files/14436976/gchp.20190701_0000z.log)

**Comment by lizziel** [comment]
This is interesting! You seem to be encountering a 14.3 bug that went under the radar before. [This section of code](https://github.com/geoschem/geos-chem/blob/b288af1563a9a49a154a93bdd0f88be1f58f3f2f/GeosCore/input_mod.F90#L2909-L2927) is written to write out with format 100, which is defined as logicals. But the actual variable at that line is real not logical. Hence the variable-type mismatch error you are seeing. Try changing this line:
`WRITE( 6, 100 ) 'CO2 VMR in ppmv             : ', Input_Opt%RRTMG_CO2_ppmv`
To this instead (format 115 instead of 100):
`WRITE( 6, 115 ) 'CO2 VMR in ppmv             : ', Input_Opt%RRTMG_CO2_ppmv`

**Comment by yuanjianz** [comment]
Hi @lizziel, I tried recompiling under Debug after editting the line you suggested. I also turned on verbose output in geoschem_config.yml. Here are the error reports:
```
===============================================================================
DO_FULLCHEM : Interface between GEOS-Chem and the solver

* Sulfur sea salt chemistry is computed in KPP
* Sulfur in-cloud chemistry is computed in KPP
* Photolysis is activated -- rates computed by Cloud-J
===============================================================================
     ### Species Unit Conversion: kg -> molec/cm3 ###
     ### Do_FullChem: after computing J-values

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   RANK 0 PID 8 RUNNING AT compute1-exec-15.ris.wustl.edu
=   KILLED BY SIGNAL: 9 (Killed)
===================================================================================
```
```
forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source             
gchp               000000000AEA093A  Unknown               Unknown  Unknown
libpthread-2.17.s  00007FA4300E7630  Unknown               Unknown  Unknown
gchp               00000000071FC077  fjx_sub_mod_mp_bl        1331  fjx_sub_mod.f90
gchp               00000000071FB5C8  fjx_sub_mod_mp_mi        1303  fjx_sub_mod.f90
gchp               00000000071F3BB6  fjx_sub_mod_mp_op        1214  fjx_sub_mod.f90
gchp               00000000071B14AD  fjx_sub_mod_mp_ph         572  fjx_sub_mod.f90
gchp               0000000007147841  cld_sub_mod_mp_cl         216  cld_sub_mod.f90
```
So it seems ture that the problem is with Cloud-J module in `fjx_sub_mod_mp_blkslv` subroutine. I am not familiar with this newly updated submodule. Do you have any clue about this?

[lsf-711754.txt.log](https://github.com/geoschem/GCHP/files/14440808/lsf-711754.txt.log)
[gchp.20190701_0000z.log](https://github.com/geoschem/GCHP/files/14440809/gchp.20190701_0000z.log)

**Comment by lizziel** [comment]
Could you try building the Cloud-J standalone model to see if you get the same error? Instructions for download, building, and running are on the README at https://github.com/geoschem/Cloud-J/. You should use the main branch for this and configure with `-DCMAKE_BUILD_TYPE=Debug`.

**Comment by yuanjianz** [comment]
Hi @lizziel , sure! I attached the log files below. Interestingly, intel and gnu compiler both fails for some reasons. Intel compiler seems to fail at the BLKSLV subroutine like mentioned above, while GNU fails indicating a erroneous math operation. I also tried to use gnu compiler under release mode and it halts at the same place as intel, which is weird because I successfully ran the gnu version at low resolution in GCHP.

[cloud.gnu.log](https://github.com/geoschem/GCHP/files/14452802/cloud.gnu.log)
[cloud.intel.log](https://github.com/geoschem/GCHP/files/14452803/cloud.intel.log)

**Comment by lizziel** [comment]
I will close this issue since anyone else running into it should see Cloud-J issue https://github.com/geoschem/Cloud-J/issues/8.

**Comment by yuanjianz** [comment]
I figured out the problem with Cloud-J within GCHP after doing some tests on both standalone and GCHP built-in versions. 
> Looking at the common loop problem: fjx_sub_mod.f90:1355 It may be that too many 4D arrays in the FP list may be troublesome That section (the call GEN_ID) could be rewritten with simple 3D transfer arrays with a little extra code, abut it might be simpler?

BLKSLV subroutine in fjx_sub_mod.f90 utilizes a considerable amount of 4D arrays, which requires larger stack memory limits for intel compiler. Otherwise it would arouse SegFault on servers that have a low limit. In standalone version of Cloud-J, this can be fixed by simply increase shell limit `ulimit -s unlimited`. 

However, since OMP is often disabled, GCHP jobs usually rely more on MPI parallelization and are insensitve to stack memory limits (comparing to GEOS-Chem Classic). The merging of Cloud-J might cause some issues if users are not suggested to increase their stack memory limits in the documentation.

On the other hand, GCHP MPI jobs under schedulers might face stack memory limit managing issue as well. On Compute1 at WashU, we use LSF scheduler to execute `mpiexec -n $cores ./gchp` in a run script. However, LSF will not pass shell limit settings to MPI component automatically even if declared in the run script. Shell limit declaration needs to be right after the `mpiexec` command. My current workaround is using `mpiexec -n $cores ./execute.sh` instead.
```
# execute.sh
ulimit -s unlimited
./gchp
```
To summarize, intel compiled GCHP 14.3.0 might run into a SegFault if not mannually raising the stack memory limit in each MPI component.

**Comment by lizziel** [comment]
Thanks for following up on this. I updated the Cloud-J README with this suggestion for cloud-j standalone.
